src/cmake-tools.ts

```diff
@@ -265,26 +266,42 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
     let drv: CMakeDriver;
     const preferredGenerators = this.getPreferredGenerators();
     const preConditionHandler = async (e: CMakePreconditionProblems) => this.cmakePreConditionProblemHandler(e);
-    if (this.workspaceContext.config.useCMakeServer) {
-      if (cmake.isServerModeSupported) {
-        drv = await CMakeServerClientDriver
-                  .create(cmake, this.workspaceContext.config, kit, workspace, preConditionHandler, preferredGenerators);
+    let communicationMode = this.workspaceContext.config.cmakeCommunicationMode;
+
+    if (communicationMode == 'automatic') {
+      if (cmake.isFileApiModSupported) {
+        communicationMode = 'fileAPI';
+      } else if (cmake.isServerModeSupported) {
+        communicationMode = 'serverAPI';
       } else {
+        communicationMode = 'legacy';
+      }
+
+      if (communicationMode != 'fileAPI') {
         log.warning(
           localize('please.upgrade.cmake',
-            'CMake Server is not available with the current CMake executable. Please upgrade to CMake {0} or newer.',
-            versionToString(cmake.minimalServerModeVersion)));
-        drv = await LegacyCMakeDriver
-                  .create(cmake, this.workspaceContext.config, kit, workspace, preConditionHandler, preferredGenerators);
+            'CMake FileApi is not available with the current CMake executable. Please upgrade to CMake {0} or newer.',
```

Comment:
I think the purpose of this message was related to the fact that "legacy" mode provides little actual CMake support.  You can still get decent support without file api, so people might complain about this message.  We could probably show a similar message if neither server mode nor file api are available.

"For the best experience, CMake server or file-api support is required.  Please upgrade CMake to {0} or newer." (serverModeVersion)